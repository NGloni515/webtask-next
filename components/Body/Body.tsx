import React, { useEffect, useState } from 'react';
import {
    Container,
    MainImage,
    CarGeneralData,
    CarContainer,
    ContainerTitle,
    ContainerParam,
    ContainerImportantData,
    ParamsContainer,
    ContainerShare,
    CountNumber,
    ImageElement,
    PerformanceCard,
    PerformanceTitle,
    MainContainer,
    MobileButton,
    MobileFooter,
    MobileFooterTitle,
    MobileFooterDescription,
    Text
  } from './BodyElements';
  import {
    IoMail
  } from 'react-icons/io5';

import "react-responsive-carousel/lib/styles/carousel.min.css"; 
import { Carousel } from 'react-responsive-carousel';
import { EMPTY_CAR, Car } from '../../models/car.model';
import axios from 'axios';

const Body = () => {

    const [car, setCar] = useState<Car>(EMPTY_CAR);
    const [currentImage, setCurrentImage] = useState<string>("");

    useEffect(() => {
        const myAsyncFunc = async() => {
            const response :Car = (await axios.get(`/api/cars`)).data.car;
            setCar(response)
            setCurrentImage(response.imagesUrls[0])
        }
        myAsyncFunc();
    }, []);
    
    const {name, year, priceMin, priceMax, mileage, itemNumber, vin, views, saves, shares, extrior, performance, imagesUrls} = car;
    return(
        <div className="grid-container">
            <div className="desktop-hide">
                <Carousel className="carousel-style" showArrows={false} showStatus={false} showThumbs={false} >
                    {imagesUrls && imagesUrls.map((element, index) => 
                        <div key={index} className="slider-item-div">
                            <img src={element} alt={"car image "+ index}/>
                        </div>
                    )}
                </Carousel>
            </div>
            <MainContainer >
                <MainImage  className="mobile-hide" src={currentImage && currentImage} />
                <CarGeneralData className="mobile-general">
                    <CarContainer className="car-container-mobile">
                        <ContainerTitle> {name} </ContainerTitle>
                        <Container className="car-container-grid">
                            <ParamsContainer className="car-container-col1">
                                <ContainerParam>Year</ContainerParam>
                                <ContainerImportantData>{year}</ContainerImportantData>
                                <ContainerParam>Price Range</ContainerParam>
                                <ContainerImportantData>${priceMin} - ${priceMax}</ContainerImportantData>
                                <ContainerParam>Mileage</ContainerParam>
                                <ContainerImportantData>{mileage} miles</ContainerImportantData>
                            </ParamsContainer>
                            <Container className="car-container-col2">
                                <ContainerParam>Item Number: {itemNumber}</ContainerParam>
                                <ContainerParam>VIN: {vin}</ContainerParam>
                                <ContainerShare>Share this car <IoMail className="icon-text" /> </ContainerShare>
                            </Container>
                            <Container className="align-center car-container-col3">
                                <Container>
                                    <ContainerParam className="no-margin">Views</ContainerParam>
                                    <CountNumber className="views">{views}</CountNumber>
                                </Container>
                                <Container>
                                    <ContainerParam className="no-margin mobile-hide">Saves</ContainerParam>
                                    <CountNumber className="mobile-hide">{saves}</CountNumber>
                                </Container>
                                <Container>
                                    <ContainerParam className="no-margin mobile-hide">Shares</ContainerParam>
                                    <CountNumber className="mobile-hide">{shares}</CountNumber>
                                </Container>
                            </Container>
                        </Container>
                    </CarContainer>
                </CarGeneralData>
            </MainContainer>
            <Container className="margin-t4 grid-image-container mobile-hide">
                {imagesUrls && imagesUrls.map((element, index) => 
                    <ImageElement key={index} className="grid-image-item" src={element} alt={"car image "+ index} onClick={()=>{setCurrentImage(element)}}/>
                )}
            </Container>
            <MobileButton className="desktop-hide">
                CALL US
            </MobileButton>
            <Container className="grid-performance-container">
                <PerformanceCard className="performance-card-responsive">
                    <PerformanceTitle>EXTRIOR</PerformanceTitle>
                    <Container className="grid-table">
                         <Text className="grid-table-element-right">Cylinders</Text> <Text className="grid-table-element-left">{extrior && extrior.cylinders}</Text>
                         <Text className="grid-table-element-right">City MPG</Text> <Text className="grid-table-element-left">{extrior && extrior.cityMpg} MPG</Text>
                         <Text className="grid-table-element-right">Highway MPG</Text> <Text className="grid-table-element-left">{extrior && extrior.highMpg} MPG</Text>
                         <Text className="grid-table-element-right">Engine</Text> <Text className="grid-table-element-left">{extrior && extrior.engine}</Text>
                    </Container>
                </PerformanceCard>
                <PerformanceCard className="performance-card-responsive">
                    <PerformanceTitle>PERFORMANCE</PerformanceTitle>
                    <Container className="grid-table">
                        <Text className="grid-table-element-right">Cylinders</Text> <Text className="grid-table-element-left">{performance && performance.cylinders}</Text>
                        <Text className="grid-table-element-right">City MPG</Text> <Text className="grid-table-element-left">{performance && performance.cityMpg} MPG</Text>
                        <Text className="grid-table-element-right">Highway MPG</Text> <Text className="grid-table-element-left">{performance && performance.highMpg} MPG</Text>
                        <Text className="grid-table-element-right">Engine</Text> <Text className="grid-table-element-left">{performance && performance.engine}</Text>
                    </Container>
                </PerformanceCard>
            </Container>
            <MobileFooter className="desktop-hide">
                <MobileFooterTitle>
                    About Us / Terms / Privacy Policy
                </MobileFooterTitle>
                <MobileFooterDescription>
                    PLS is a registered service mark and other marks are service marks of PLS Financial Services, Inc 2016
                </MobileFooterDescription>
            </MobileFooter>
        </div>
        
        
    )
}

export default Body;