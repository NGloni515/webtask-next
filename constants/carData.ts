import { Car } from "../models/car.model";

export const CAR_1: Car = {
    "name": "Ford Focus", 
    "year": 2012,
    "priceMin": 8500,
    "priceMax": 9000,
    "mileage": 200000,
    "itemNumber": 1395,
    "vin": "3GNDA13D96S631406",
    "views": 37,
    "saves": 20,
    "shares": 15,
    "extrior": {
        "cylinders": 4,
        "cityMpg": 20,
        "highMpg": 20,
        "engine": 1.3
    },
    "performance": {
        "cylinders": 4,
        "cityMpg": 20,
        "highMpg": 20,
        "engine": 1.3
    },
    "imagesUrls":[
        "https://i.pinimg.com/originals/c6/97/15/c697157b77bca2e6d8449ad917b61924.jpg",
        "https://noticias.coches.com/wp-content/uploads/2014/02/Ford-Focus-2014-interior-04-650x419.jpg",
        "https://i.blogs.es/367759/presentacion-ford-focus-20/450_1000.jpg",
        "https://carnovo-wordpress-media.s3.eu-west-1.amazonaws.com/wp-content/uploads/2018/04/10151522/focus-2.jpg",
        "https://i0.wp.com/geeksroom.com/wp-content/uploads/2018/04/ford-focus-00.jpg?resize=680%2C384&ssl=1",
        "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcS-w7l7FSKQ3RWVjG3B2_5xjqkO70N5E2zFCg&usqp=CAU"
    ]
};