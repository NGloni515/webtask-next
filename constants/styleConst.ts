export const COLORS = {
    LIGHT_GREEN: '#3fb34a',
    GREEN: '#106838',
    LIGHT_GRAY: '#efeff2',
    WHITE: '#ffffff',
    BLACK: '#000000',
    SOFT_BLACK: '#343434',
    GRAY: '#9b9b9b',
    RED: '#ec4f38',
    ORANGE: '#f88200',
}

export const FONTS = {
    MAIN: 'Helvetica Neue'
}
