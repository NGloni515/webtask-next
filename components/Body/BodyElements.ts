import styled from 'styled-components';
import { COLORS, FONTS } from '../../constants/styleConst';

export const ContainerTitle = styled.h4`
    font-family: ${FONTS.MAIN};
    font-size: 2rem;
    font-weight: bold;
    margin: 0;
    padding: 0;
    text-align: left;
`;

export const ContainerParam = styled.h4`
    font-family: ${FONTS.MAIN};
    font-size: 0.7rem;
    font-weight: normal;
    margin: 0 0 8px 0;
    padding: 0;
    text-align: left;
    color: ${COLORS.GRAY};
`;

export const ContainerImportantData = styled.h4`
    font-family: ${FONTS.MAIN};
    font-size: 0.8rem;
    font-weight: bold;
    margin: 0 0 16px 0;
    padding: 0;
    text-align: left;
    color: ${COLORS.BLACK};
`;

export const ContainerShare = styled.h4`
    font-family: ${FONTS.MAIN};
    font-size: 0.7rem;
    font-weight: bold;
    margin: 24px 0;
    padding: 0;
    text-align: left;
    color: ${COLORS.SOFT_BLACK};
`;

export const CountNumber = styled.h4`
    font-size: 0.8rem;
    font-weight: bold;
    margin: 0;
    padding: 0;
    text-align: center;
    color: ${COLORS.LIGHT_GREEN};
`;

export const VerticalElement = styled.h4`
    font-family: ${FONTS.MAIN};
    font-size: 0.7rem;
    font-weight: bold;
    margin: 24px 0;
    padding: 0;
    text-align: left;
    color: ${COLORS.SOFT_BLACK};
`;

export const ParamsContainer =  styled.div`
  margin: 32px 0;
  padding: 0;
`;

export const Container =  styled.div`
  margin: 0;
  padding: 0;
`;

export const MainContainer =  styled.div`
  margin: 0;
  padding: 0;
  width: 100vw;
`;

export const MainImage =  styled.img`
  margin: 0;
  padding: 0;
  width: 71vw;
  max-width: 71vw;
  min-width: 71vw;
  height: 432px;
  float: left;
  object-fit: cover;
`;

export const ImageElement =  styled.img`
  width: auto;
  max-width: 16vw;
  height: 170px;
  object-fit: cover;
`;


export const CarGeneralData =  styled.div`
  padding: 0;
  margin: 0 0 0 0;
  width: 29vw;
  max-width: 29vw;
  min-width: 29vw;
  height: 432px;
  float: right;
  background-color: ${COLORS.WHITE};
`;

export const CarContainer =  styled.div`
  padding: 0;
  margin: 48px 0 0 40px;
`;

export const PerformanceCard =  styled.div`
  padding: 40px;
  margin: 0;
  max-width: 39vw;
  min-width: 39vw;
  height: 266px;
  background-color: ${COLORS.WHITE};
`;

export const PerformanceTitle =  styled.div`
  font-family: ${FONTS.MAIN};
  font-size: 1.2rem;
  font-weight: bold;
  margin: 0 0 8px 0;
  padding: 0;
  text-align: left;
  color: ${COLORS.GRAY};
`;

// MOBILE ONLY ELEMENTS

export const ImageSlider =  styled.div`
  width: 100vw;
  height: 256px;
  position: relative;
  margin: auto;
`;

export const MobileButton =  styled.button`
  width: 100vw;
  height: 50px;
  margin: 13px 0 5px 0;
  background-color: ${COLORS.LIGHT_GREEN};
  color: ${COLORS.WHITE};
  font-size: 1rem;
  border: none;
`;

export const MobileFooter =  styled.footer`
  width: 100vw;
  margin: 13px 0 5px 0;
  background-color: ${COLORS.BLACK};
  color: ${COLORS.WHITE};
  font-size: 0.8rem;
  text-align: center;
`;

export const MobileFooterTitle =  styled.p`
  margin: 10px 0;
`;

export const MobileFooterDescription =  styled.p`
  font-size: 0.7rem;
  margin: 15px;
`;

export const Text =  styled.p`
  margin: 0;
`;