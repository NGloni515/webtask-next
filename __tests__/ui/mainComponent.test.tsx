import { render } from '@testing-library/react'
import Body from '../../components/Body/Body'
import '@testing-library/jest-dom'
// import axios from 'axios'
import { CAR_1 } from '../../constants/carData';

describe('Body', () => {
  beforeEach(()=>{
    global.fetch = jest.fn(()=>Promise.resolve({
      json: ()=>Promise.resolve(CAR_1)
    })) as jest.Mock;
  });

  it('Match snapshot Carousel', async() => {
    const {asFragment} = render(<Body/>)
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders a Carousel', async() => {
    const {getByTestId} = render(<Body/>)
    const carousel = getByTestId("grid-container")
    expect(carousel).toBeInTheDocument()
  })

  it('renders a Carousel', async() => {
    const {getByTestId} = render(<Body/>)
    const imgContainer = getByTestId("image-container")
    expect(imgContainer).toBeInTheDocument()
  })
})