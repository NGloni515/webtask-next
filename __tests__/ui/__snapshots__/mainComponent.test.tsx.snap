// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Body Match snapshot Carousel 1`] = `
<DocumentFragment>
  <div
    class="grid-container"
    data-testid="grid-container"
  >
    <div
      class="desktop-hide"
    />
    <div
      class="BodyElements__MainContainer-sc-cbg9h8-8 hfgNCg"
    >
      <img
        class="BodyElements__MainImage-sc-cbg9h8-9 kVUcyO mobile-hide"
        src=""
      />
      <div
        class="BodyElements__CarGeneralData-sc-cbg9h8-11 kgDEDz mobile-general"
      >
        <div
          class="BodyElements__CarContainer-sc-cbg9h8-12 igZHKP car-container-mobile"
        >
          <h4
            class="BodyElements__ContainerTitle-sc-cbg9h8-0 dTTkrn"
          >
              
          </h4>
          <div
            class="BodyElements__Container-sc-cbg9h8-7 hyCVVz car-container-grid"
          >
            <div
              class="BodyElements__ParamsContainer-sc-cbg9h8-6 jVRgAz car-container-col1"
            >
              <h4
                class="BodyElements__ContainerParam-sc-cbg9h8-1 hygXcz"
              >
                Year
              </h4>
              <h4
                class="BodyElements__ContainerImportantData-sc-cbg9h8-2 cyKKaw"
              >
                0
              </h4>
              <h4
                class="BodyElements__ContainerParam-sc-cbg9h8-1 hygXcz"
              >
                Price Range
              </h4>
              <h4
                class="BodyElements__ContainerImportantData-sc-cbg9h8-2 cyKKaw"
              >
                $0 - $0
              </h4>
              <h4
                class="BodyElements__ContainerParam-sc-cbg9h8-1 hygXcz"
              >
                Mileage
              </h4>
              <h4
                class="BodyElements__ContainerImportantData-sc-cbg9h8-2 cyKKaw"
              >
                0 miles
              </h4>
            </div>
            <div
              class="BodyElements__Container-sc-cbg9h8-7 hyCVVz car-container-col2"
            >
              <h4
                class="BodyElements__ContainerParam-sc-cbg9h8-1 hygXcz"
              >
                Item Number: 0
              </h4>
              <h4
                class="BodyElements__ContainerParam-sc-cbg9h8-1 hygXcz"
              >
                VIN: 
              </h4>
              <h4
                class="BodyElements__ContainerShare-sc-cbg9h8-3 bpkRSf"
              >
                Share this car 
                <svg
                  class="icon-text"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 512 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M424 80H88a56.06 56.06 0 00-56 56v240a56.06 56.06 0 0056 56h336a56.06 56.06 0 0056-56V136a56.06 56.06 0 00-56-56zm-14.18 92.63l-144 112a16 16 0 01-19.64 0l-144-112a16 16 0 1119.64-25.26L256 251.73l134.18-104.36a16 16 0 0119.64 25.26z"
                  />
                </svg>
                 
              </h4>
            </div>
            <div
              class="BodyElements__Container-sc-cbg9h8-7 hyCVVz align-center car-container-col3"
            >
              <div
                class="BodyElements__Container-sc-cbg9h8-7 hyCVVz"
              >
                <h4
                  class="BodyElements__ContainerParam-sc-cbg9h8-1 hygXcz no-margin"
                >
                  Views
                </h4>
                <h4
                  class="BodyElements__CountNumber-sc-cbg9h8-4 fRqoPJ views"
                >
                  0
                </h4>
              </div>
              <div
                class="BodyElements__Container-sc-cbg9h8-7 hyCVVz"
              >
                <h4
                  class="BodyElements__ContainerParam-sc-cbg9h8-1 hygXcz no-margin mobile-hide"
                >
                  Saves
                </h4>
                <h4
                  class="BodyElements__CountNumber-sc-cbg9h8-4 fRqoPJ mobile-hide"
                >
                  0
                </h4>
              </div>
              <div
                class="BodyElements__Container-sc-cbg9h8-7 hyCVVz"
              >
                <h4
                  class="BodyElements__ContainerParam-sc-cbg9h8-1 hygXcz no-margin mobile-hide"
                >
                  Shares
                </h4>
                <h4
                  class="BodyElements__CountNumber-sc-cbg9h8-4 fRqoPJ mobile-hide"
                >
                  0
                </h4>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="BodyElements__Container-sc-cbg9h8-7 hyCVVz margin-t4 grid-image-container mobile-hide"
      data-testid="image-container"
    />
    <button
      class="BodyElements__MobileButton-sc-cbg9h8-16 eTcPPZ desktop-hide"
    >
      CALL US
    </button>
    <div
      class="BodyElements__Container-sc-cbg9h8-7 hyCVVz grid-performance-container"
    >
      <div
        class="BodyElements__PerformanceCard-sc-cbg9h8-13 fnisDo performance-card-responsive"
      >
        <div
          class="BodyElements__PerformanceTitle-sc-cbg9h8-14 jRvUoV"
        >
          EXTRIOR
        </div>
        <div
          class="BodyElements__Container-sc-cbg9h8-7 hyCVVz grid-table"
        >
          <p
            class="BodyElements__Text-sc-cbg9h8-20 eMxcTw grid-table-element-right"
          >
            Cylinders
          </p>
           
          <p
            class="BodyElements__Text-sc-cbg9h8-20 eMxcTw grid-table-element-left"
          >
            0
          </p>
          <p
            class="BodyElements__Text-sc-cbg9h8-20 eMxcTw grid-table-element-right"
          >
            City MPG
          </p>
           
          <p
            class="BodyElements__Text-sc-cbg9h8-20 eMxcTw grid-table-element-left"
          >
            0 MPG
          </p>
          <p
            class="BodyElements__Text-sc-cbg9h8-20 eMxcTw grid-table-element-right"
          >
            Highway MPG
          </p>
           
          <p
            class="BodyElements__Text-sc-cbg9h8-20 eMxcTw grid-table-element-left"
          >
            0 MPG
          </p>
          <p
            class="BodyElements__Text-sc-cbg9h8-20 eMxcTw grid-table-element-right"
          >
            Engine
          </p>
           
          <p
            class="BodyElements__Text-sc-cbg9h8-20 eMxcTw grid-table-element-left"
          >
            0
          </p>
        </div>
      </div>
      <div
        class="BodyElements__PerformanceCard-sc-cbg9h8-13 fnisDo performance-card-responsive"
      >
        <div
          class="BodyElements__PerformanceTitle-sc-cbg9h8-14 jRvUoV"
        >
          PERFORMANCE
        </div>
        <div
          class="BodyElements__Container-sc-cbg9h8-7 hyCVVz grid-table"
        >
          <p
            class="BodyElements__Text-sc-cbg9h8-20 eMxcTw grid-table-element-right"
          >
            Cylinders
          </p>
           
          <p
            class="BodyElements__Text-sc-cbg9h8-20 eMxcTw grid-table-element-left"
          >
            0
          </p>
          <p
            class="BodyElements__Text-sc-cbg9h8-20 eMxcTw grid-table-element-right"
          >
            City MPG
          </p>
           
          <p
            class="BodyElements__Text-sc-cbg9h8-20 eMxcTw grid-table-element-left"
          >
            0 MPG
          </p>
          <p
            class="BodyElements__Text-sc-cbg9h8-20 eMxcTw grid-table-element-right"
          >
            Highway MPG
          </p>
           
          <p
            class="BodyElements__Text-sc-cbg9h8-20 eMxcTw grid-table-element-left"
          >
            0 MPG
          </p>
          <p
            class="BodyElements__Text-sc-cbg9h8-20 eMxcTw grid-table-element-right"
          >
            Engine
          </p>
           
          <p
            class="BodyElements__Text-sc-cbg9h8-20 eMxcTw grid-table-element-left"
          >
            0
          </p>
        </div>
      </div>
    </div>
    <footer
      class="BodyElements__MobileFooter-sc-cbg9h8-17 ghIHSa desktop-hide"
    >
      <p
        class="BodyElements__MobileFooterTitle-sc-cbg9h8-18 jyPmCW"
      >
        About Us / Terms / Privacy Policy
      </p>
      <p
        class="BodyElements__MobileFooterDescription-sc-cbg9h8-19 cJluul"
      >
        PLS is a registered service mark and other marks are service marks of PLS Financial Services, Inc 2016
      </p>
    </footer>
  </div>
</DocumentFragment>
`;
